### PR TITLE
Allow specifying tags for aten operators in native_functions.yaml

### PR DESCRIPTION
**Pull request by anjali411**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72549

This PR:
1. Adds a new yaml file. For each new tag, we will require an entry in this file along with a mandatory description.
2. There are checks to ensure that any tags added for native_functions are valid tags (i.e., they exist in tags.yaml)
3. Each NativeFunction object holds tags as a set of strings (this logic can be found in model.py)

Most file changes are to update the call to `parse_native_yaml` to include the path to `native_functions.yaml` as well as `tags.yaml`

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -730,7 +730,8 @@
   variants: function, method
   device_check: NoCheck
   device_guard: False
-  tags: inplace_view
+  tags:
+    is_inplace_view: True
```
Comment (ezyang) [inline aten/src/ATen/native/native_functions.yaml:734]: This format seems... worse? Do we have use cases for non-boolean valued tags?

**Comment by ezyang** [comment]
oops, didn't see not for review. Unsubscribing for now, please re-cc me when ready

File: tools/autograd/gen_autograd.py [hunk 2]
```diff
@@ -41,28 +41,30 @@
...
     out: str,
     autograd_dir: str,
     operator_selector: SelectiveBuilder,
     disable_autograd: bool = False,
 ) -> None:
     # Parse and load derivatives.yaml
     differentiability_infos = load_derivatives(
-        os.path.join(autograd_dir, 'derivatives.yaml'), native_functions_path)
+        os.path.join(autograd_dir, 'derivatives.yaml'), native_functions_path,
+        tags_path)
```
Comment (ezyang) [inline tools/autograd/gen_autograd.py:53]: this probably fails flake

**Comment by ezyang** [comment]
These tags still aren't exposed from Python, is that right? It would be nice to parse in the descriptions so that we can then generate appropriate `__doc__` on the Python side representation (seems better than slinging strings around in Python).

File: tools/codegen/gen.py [hunk 3]
```diff
@@ -144,12 +145,38 @@ def parse_native_yaml_struct(es: object, path: str = "<stdin>") -> ParsedYaml:
...
+            assert desc != ''
+            rs.add(name)
+    return rs
+
+@functools.lru_cache(maxsize=None)
+def parse_tags_yaml(path: str) -> Set[str]:
+    # TODO: parse tags.yaml and create a tags database (a dict of tag name mapping to a Tag object)
+    with open(path, 'r') as f:
+        es = yaml.load(f, Loader=LineLoader)
+        ValidTags = parse_tags_yaml_struct(es, path=path)
```
Comment (ezyang) [inline tools/codegen/gen.py:169]: nit: upper-cased identifier here isn't very idiomatic

**Review by ezyang — APPROVED** [review]

**Comment by anjali411** [comment]
> ValidTags

no they aren't yet but the next PR in python

> These tags still aren't exposed from Python, is that right? It would be nice to parse in the descriptions so that we can then generate appropriate `__doc__` on the Python side representation (seems better than slinging strings around in Python).

No they aren't yet but the next PR in the stack exposes the tags through the torch.ops API.

File: BUILD.bazel [hunk 4]
```diff
@@ -185,6 +185,7 @@ genrule(
     name = "all_generated_code",
     srcs = [
         "aten/src/ATen/native/native_functions.yaml",
+        "aten/src/ATen/native/tags.yaml"
```
Comment (malfet) [inline BUILD.bazel:188]: I guess that was the source of the problem :)
```suggestion
        "aten/src/ATen/native/tags.yaml",
```

**Comment by malfet** [comment]
Reverting as this change broke internal build systems. Significant build system changes still have to go thru the old land path

**Pull request closed without merging**